0025-interface-for-circuit-operations.md

```diff
@@ -0,0 +1,181 @@
+  to this object at run-time (notably _not_ the parameters themselves).  This is
+  to enable classical resource tracking and type safety.
+
+- `state`: an opaque object that contains the full state of the object, which
+  can be passed to some as-yet-undefined method to reconstruct the object.  This
+  is for serialisation.
+
+Some things `Operation` should not do:
+
+- force objects to make decisions about their synthesis
+- force all objects to be stateful (i.e. there should be a valid implementation
+  of `HGate` that is a singleton)
+- require objects to construct / be able to construct a `QuantumCircuit` through
+  any of the abstract methods
+
+These things mean that the following properties or methods should _not_ be part
+of `Operation`:
+
+- `condition`: this should be part of separate classical control flow, and its
+  stateful nature is one of the current reasons for a lot of copying of
+  `Instruction` instances.
+
+- `duration` and `unit`: these are properties of the output of the transpiler,
+  and should also depend on the particular qubits in use.  This likely wants to
+  become part of `CircuitInstruction` (see Qiskit/qiskit-terra#8093).
+
+- `assemble`: this is particular to the conversion to the `Qobj` type, which may
+  not exist in the long-term, but even in the short-term, this is conversion
+  behaviour and the interfaces should be well-enough defined such that this does
+  not need to be specialised.
+
+- `definition` or `_define`: this ties synthesis to the object, causing us to
+  carry around extra circuits (and frequently copy them).
+
+
+## Alternative Approaches
+
+It is in theory possible to maintain the status quo, and have `Instruction` be
+the definition of the transpiler-input interface.  Some things (such as
+pluggable, controllable synthesis passes) are still possible to do in this
+structure.  The memory problems are (at best) exacerbated by `Instruction`,
+though, and it is difficult to fix this will maintaining backwards
+compatibility.  This also would do nothing to address the general design
+problems with operations currently needing to define their own synthesis.
+
+The memory usage is already showing problematic strain for scalability.  See:
+
+- Qiskit/qiskit-terra#7485
+- Qiskit/qiskit-terra#6991
+- Qiskit/qiskit-terra#5895
+
+Having all operations have to define their own `definition` causes cyclic
+dependency and hierarchy issues: `quantum_info` classes generally should not
+need to care about the circuit model, since they are using a different
+formalism.  This also means that objects with more than one synthesis method
+have to define a "preferred" synthesis method.  This is an issue for things like
+the multi-controlled X.
+
+
+## Questions
+
+1. At this point, the exact requirements for `Operation` are not fully
+   finalised.  This RFC will not be ready for implementation until they are.
+
+2. Will classical operations also be `Operation`, or will we do something
+   different with them?
```

Comment:
For me I think it makes the most sense to have classical operations be based on the same base class. We can specialize on top of it if there are stricter requirements, but having a common class for all operations makes working with circuits internally a lot easier I think. Is thing a operation, if so we can use it in a circuit? Then for things consuming circuits at a low level (like qpy) know how to work with the base element in the circuit.